In [2]:
import pandas as pd
import numpy as np
import os

Question 1: clean the products data

In [3]:
my_path = r'C:\Users\nkasb\Documents\Instacart Basket Analysis'

In [5]:
df_prods = pd.read_csv(os.path.join(my_path, '2. Data', 'Original Data', 'products.csv'), index_col = False)


In [6]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [7]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [8]:
df_nan


product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [10]:
df_prods_clean


product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [11]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [12]:
df_dups.shape

(5, 5)

In [13]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [14]:
df_prods_final = df_prods_clean.drop_duplicates()

In [15]:
df_prods_final.shape

(49672, 5)

Question 2: check the orders data for irregularities

In [17]:
df_ords = pd.read_csv(os.path.join(my_path, '2. Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [18]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

day-since_prior_order has 50% percentile of 7 but a mean of 11. it suggests that that column has outliers which artificially increase the mean. it max value is 30 which is largely above the 50% percentile 

Question 3: check orders data for mix values

In [19]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There is no mix value in the orders data

Question 5: check the orders data for missing values

In [20]:
df_ords_missing = df_ords[df_ords['order_id'].isnull() == True]

In [21]:
df_ords_missing


Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [22]:
df_ords_missing1 = df_ords[df_ords['user_id'].isnull() == True]

In [23]:
df_ords_missing1

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [24]:
df_ords_missing2 = df_ords[df_ords['order_number'].isnull() == True]

In [25]:
df_ords_missing2

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [26]:
df_ords_missing3 = df_ords[df_ords['order_number'].isnull() == True]

In [27]:
df_ords_missing3

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [33]:
df_ords_missing4 = df_ords[df_ords['orders_day_of_week'].isnull() == True]

In [34]:
df_ords_missing4

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [35]:
df_ords_missing4 = df_ords[df_ords['order_hour_of_day'].isnull() == True]

In [36]:
df_ords_missing4

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [37]:
df_ords_missing5 = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [38]:
df_ords_missing5

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

In [39]:
df_ords.shape

(3421083, 8)

206209 missing values in days_since_prior_order represents about 6% of the dataset

In [40]:
df_ords_missing5.describe()

Unnamed: 0      order_id        user_id  order_number  \
count  2.062090e+05  2.062090e+05  206209.000000      206209.0   
mean   1.712637e+06  1.708462e+06  103105.000000           1.0   
std    9.874816e+05  9.881299e+05   59527.555167           0.0   
min    0.000000e+00  2.000000e+01       1.000000           1.0   
25%    8.579730e+05  8.507300e+05   51553.000000           1.0   
50%    1.717068e+06  1.706246e+06  103105.000000           1.0   
75%    2.570518e+06  2.564292e+06  154657.000000           1.0   
max    3.421069e+06  3.421081e+06  206209.000000           1.0   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count       206209.000000      206209.000000                     0.0  
mean             2.754118          13.626597                     NaN  
std              2.076205           4.223769                     NaN  
min              0.000000           0.000000                     NaN  
25%              1.000000          11.000000                     NaN  
50%              3.000000          14.000000                     NaN  
75%              5.000000          17.000000                     NaN  
max              6.000000          23.000000                     NaN

all users with NaN days_since_prior-order has order_number 1 mean it was their first order. In that case it is not possible
to calculate the number of days since the previous order. To address these missing values, i will replace the NAN by 0.

In [41]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [42]:
df_ords_missing6 = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [43]:
df_ords_missing6

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [44]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [45]:
df_ords_dups


Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

there is no duplicates.

In [46]:
df_ords.to_csv(os.path.join(my_path, '2. Data','Prepared Data', 'orders_clean_final.csv'))

In [47]:
df_prods.to_csv(os.path.join(my_path, '2. Data','Prepared Data', 'products_clean_final.csv'))